[quant][graphmode][fx][fp16] Add fp16 support for layer_norm

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52865 [quant][graphmode][fx][fp16] Add fp16 support for silu
* #52864 [quant][graphmode][fx][fp16] Add fp16 support for tanh
* #52863 [quant][graphmode][fx][fp16] Add fp16 support for sigmoid
* **#52862 [quant][graphmode][fx][fp16] Add fp16 support for layer_norm**
* #52811 [quant][graphmode][fx][fp16] Add fp16 support for torch.sum
* #52810 [quant][graphmode][fx][fp16] Add fp16 support for div
* #52809 [quant][graphmode][fx][fp16] Add fp16 support for sub

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

jerryzh168 committed:
[quant][graphmode][fx][fp16] Add fp16 support for layer_norm

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -2622,11 +2622,52 @@ def forward(self, input):
+    def _test_norm_float16_impl(
+            self, float_module, float_op, op_args, data,
+            skip_op_arg_for_functional=False):
+        ''' Test for normalization op, float_op can be torch op or functional op,
+        op_args is a list of positional argument for the module/op
+        '''
+        class M(torch.nn.Module):
+            def __init__(self, is_module):
+                super(M, self).__init__()
+                self.is_module = is_module
+                if self.is_module:
+                    self.op = float_module(*op_args)
+                else:
+                    self.op = float_op
+
+            def forward(self, input):
+                if self.is_module:
+                    return self.op(input)
+                else:
+                    args = [input]
+                    if not skip_op_arg_for_functional:
+                        args += op_args
+                    return self.op(*args)
```
optional: this seems hard to read, at least for me (it may be different for others). I wonder if there were two less complicated classes, maybe one for modules and one for functions, if it would be easier to read.  I personally would value readability more than code reuse in this context.  I'm flexible though.

jerryzh168 replied:
sure, this is a little bit hard to read, I can split it to two modules

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -725,41 +732,78 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
+        int8_dtypes = [
+            (torch.quint8, torch.qint8, None)
+        ]
+        supported_dtypes = {
```
optional for a future PR, would it make sense for these mappings to live somewhere else in a mapping only file?

jerryzh168 replied:
yeah, I plan to refactor this part a bit later, maybe we can put the type tuple in a function. for mappings I think it would make sense to keep them here for now since they are related to the patterns above (in register_patterns), but maybe we can include them in the patterns, will need to think about it more. let me know if you have any suggestions.

vkuzo reviewed (approved):
lg!

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx][fp16] Add fp16 support for layer_norm"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_layer_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26672272](https://our.internmc.facebook.com/intern/diff/D26672272)

[ghstack-poisoned]
